tests/performance_tests/performance_test.py

```diff
@@ -11,203 +12,441 @@
+                maxFound = True
+            else:
+                searchTarget = searchTarget - step
+
+            scenarioResult.basicTestResult = ptbResult
+            searchResults.append(scenarioResult)
+            if not self.ptConfig.quiet:
+                print(f"searchResult: {searchTarget} : {searchResults[-1]}")
+
+        return PerformanceTest.TpsTestResult.PerfTestSearchResults(maxTpsAchieved=maxTpsAchieved, searchResults=searchResults, maxTpsReport=maxTpsReport)
+
+    def evaluateSuccess(self, test: PerformanceTestBasic, testSuccessful: bool, result: PerfTestSearchIndivResult.PerfTestBasicResult) -> bool:
+        result.targetTPS = test.ptbConfig.targetTps
+        result.expectedTxns = test.ptbConfig.expectedTransactionsSent
+        reportDict = test.report
+        result.testStart = reportDict["testStart"]
+        result.testEnd = reportDict["testFinish"]
+        result.resultAvgTps = reportDict["Analysis"]["TPS"]["avg"]
+        result.resultTxns = reportDict["Analysis"]["TrxLatency"]["samples"]
+        print(f"targetTPS: {result.targetTPS} expectedTxns: {result.expectedTxns} resultAvgTps: {result.resultAvgTps} resultTxns: {result.resultTxns}")
+
+        result.tpsExpectMet = True if result.resultAvgTps >= result.targetTPS else abs(result.targetTPS - result.resultAvgTps) < 100
+        result.trxExpectMet = result.expectedTxns == result.resultTxns
+        result.basicTestSuccess = testSuccessful
+        result.testAnalysisBlockCnt = reportDict["Analysis"]["BlocksGuide"]["testAnalysisBlockCnt"]
+        result.logsDir = test.loggingConfig.logDirPath
+
+        print(f"basicTestSuccess: {result.basicTestSuccess} tpsExpectationMet: {result.tpsExpectMet} trxExpectationMet: {result.trxExpectMet}")
+
+        return result.basicTestSuccess and result.tpsExpectMet and result.trxExpectMet
+
+    class PluginThreadOpt(Enum):
+        PRODUCER = "producer"
+        CHAIN = "chain"
+        NET = "net"
+
+    class PluginThreadOptRunType(Enum):
+        FULL = 1
+        LOCAL_MAX = 2
+
+    @dataclass
+    class PluginThreadOptResult:
+        recommendedThreadCount: int = 0
+        threadToMaxTpsDict: dict = field(default_factory=dict)
+        analysisStart: datetime = datetime.utcnow()
+        analysisFinish: datetime = datetime.utcnow()
+
+    def optimizePluginThreadCount(self,  optPlugin: PluginThreadOpt, optType: PluginThreadOptRunType=PluginThreadOptRunType.LOCAL_MAX,
+                                  minThreadCount: int=2, maxThreadCount: int=os.cpu_count()) -> PluginThreadOptResult:
+
+        resultsFile = f"{self.loggingConfig.pluginThreadOptLogsDirPath}/{optPlugin.value}ThreadResults.txt"
+
+        threadToMaxTpsDict: dict = {}
+
+        clusterConfig = copy.deepcopy(self.clusterConfig)
+        analysisStart = datetime.utcnow()
+
+        with open(resultsFile, 'w') as log:
+            log.write(f"{optPlugin.value}Threads, maxTpsAchieved\n")
+        log.close()
```

Comment:
This is redundant, and I'm a little surprised it didn't throw an exception.  The context manager `with` exists for the express purpose of calling whatever cleanup procedure an object which supports its protocol needs.

Reply:
Addressed: [33b94ba](https://github.com/AntelopeIO/leap/pull/510/commits/33b94ba2a4c25fde078bf6214689c9dbe66fb770)